File: libraryUse/ilands.cpp
#include <iostream>
#include <stack>
#include <complex>
using namespace std;

int numLand = 0;
int w, h;

int c[50][50];
using point = complex<int>;

point directions[] = {{1, 0}, {-1, 0}, {0, 1}, {0, -1},
                    {1, 1}, {-1, -1}, {-1, 1}, {1, -1},
};

point searchOne() {
  for (int i = 0; i < h; i++) {
    for (int j = 0; j < w; j++) {
      if(c[i][j] > 0) return point(j,i);
    }
  }
  cout << "this should not called!" << endl;
}
bool rangeOut(point & p) {
  int x = p.real(), y = p.imag();
  return (x < 0) || (x >= w) || (y < 0) || (y >= h);
}
void delLand(point & p) { 
  c[p.imag()][p.real()] = 0;
  numLand--;
}
bool solve(void) {
  int numIland = 0;

  cin >> w >> h;
  if(w==0 && h==0) return false;
  for (int i = 0; i < h; i++) {
    for (int j = 0; j < w; j++) {
      cin >> c[i][j];
      if (c[i][j] > 0) numLand++;
    }
  }

  stack<point> searching;
  while(numLand > 0) {
    numIland++;
    auto one = searchOne();
    searching.push(one);
    delLand(one);
    while (!searching.empty()) {
      auto one = searching.top();
      searching.pop();

      for(auto & direct1: directions) {
        auto next = one + direct1;
        if (rangeOut(next)) continue;
        if (c[next.imag()][next.real()] > 0) {
          delLand(next);
          searching.push(next);          
        }
      }
    }
  }
  cout << numIland << endl;
  return true;
}

int main(void) {
  while(solve());
  return 0;
}

File: libraryUse/Ilands.java
package libraryUse;

import java.awt.*;
import java.io.FileInputStream;
import java.io.IOException;
import java.util.Scanner;
import java.util.Stack;


public class Ilands {
    private final int w;
    private final int h;
    int[][] c;
    static Point[] directions =  {new Point(1, 0), new Point(-1, 0), new Point(0, 1), new Point(0, -1),
        new Point(1, 1), new Point(-1, -1), new Point(-1, 1), new Point(1, -1),
    };
    private int numLands;


    public Ilands(int w, int h, Scanner input) {
        this.w=w; this.h=h;
        c = new int[h][w];
        for(int i=0; i<h; i++) {
            for (int j = 0; j < w; j++) {
                c[i][j] = input.nextInt();
                if(c[i][j]>0) numLands++;
            }
        }
    }
    Point searchOne() {
        for (int i = 0; i < h; i++) {
            for (int j = 0; j < w; j++) {
                if (c[i][j] > 0) return new Point(j, i);
            }
        }
        throw new RuntimeException("should not happen!");
    }
    boolean rangeOut(Point p) {
        return p.x < 0 || p.x >= w || p.y < 0 || p.y >= h;
    }
    void delLand(Point p) {
        c[p.y][p.x] = 0;
        numLands--;
    }
    private int solve() {
        int numIlands = 0;
        Stack<Point> searching = new Stack<>();
        while(numLands>0) {
            numIlands++;
            Point first = searchOne();
            searching.push(first);
            delLand(first);
            while(!searching.empty()) {
                Point one = searching.pop();
                for(Point direct1: directions) {
                    Point next = new Point(one.x+direct1.x, one.y+direct1.y);
                    if(rangeOut(next)) continue;
                    if(c[next.y][next.x] > 0) {
                        delLand(next);
                        searching.push(next);
                    }
                }
            }
        }
        return numIlands;
    }


    public static void main(String[] args) throws IOException {
        // debug 時はファイルから読み込みのが楽
        // System.setIn(new FileInputStream("libraryUse/data4ilands/sample.in"));
        Scanner input = new Scanner(System.in);
        while(true) {
            int w = input.nextInt();
            int h = input.nextInt();
            if (w == 0 && h == 0) return;
            int result = new Ilands(w, h, input).solve();
            System.out.println(result);
        }
    }
}
